sphinx_gallery/gen_rst.py

```diff
@@ -541,15 +543,21 @@ def generate_dir_rst(
     user_index_rst = True
     if header_fname:
         user_index_rst = False
-        with codecs.open(header_fname, "r", encoding="utf-8") as fid:
+        with open(header_fname, "r", encoding="utf-8") as fid:
             header_content = fid.read()
             index_content += header_content
 
     # Add empty lines to avoid bug in issue #165
     index_content += "\n\n"
 
-    if not os.path.exists(target_dir):
-        os.makedirs(target_dir)
+    # Make all dirs ahead of time to avoid collisions in parallel processing
+    os.makedirs(target_dir, exist_ok=True)
+    image_dir = os.path.join(target_dir, "images")
+    os.makedirs(image_dir, exist_ok=True)
+    thumb_dir = os.path.join(image_dir, "thumb")
+    os.makedirs(thumb_dir, exist_ok=True)
+    if gallery_conf["jupyterlite"] is not None:
+        _add_jupyterlite_badge_logo(image_dir)
```

Comment:
These dir-making functions? I would say no because it's a fairly short block of code with no conditionals so I don't think much would be gained by moving it